### ISSUE TITLE
Failed to apply the theme as the OS rejected it!  SetSystemVisualStyle() returned 0x90070490

### ISSUE THREAD

**Issue by XayUp** [nptr/msstyleEditor#67]
Following. My pc was running the latest version of windows 8.1 and on it I created a theme using its aero.msstyles. Now I formatted my pc with this same windows 8.1 (same iso), but my modified theme no longer works on that system and when I open the modified .msstyles with msstylesEditor and test the theme, it returns the error "SetSystemVisualStyle( ) returned 0x90070490". I was confused because the theme is for the same system, and the only difference is that I formatted it, still with the same system, same iso and everything. I don't want to lose my theme like this because it was a beautiful day of work.

**Comment by chinagreenelvis** [comment]
I am having this same issue, but only if I try to create a new theme based on aero.msstyles. If I use the msstyles file from a custom theme like Simplify 10 I don't get the error and the program tests successfully.

This is a huge problem for me because I want to create a new theme starting with Aero as a base.

**Comment by nptr** [comment]
To clarify: 
- Plain aero.msstyles (from your computer, your OS) _applied via msstyleEditor_ doesn't work.
- Custom theme (like Simplify) _applied via msstyleEditor_ works.

If that is the case, would you send me the:
- original aero.msstyles
- unmodified aero.msstyles _saved with msstyleEditor_
- original custom
- unmodified custom theme _saved with msstyleEditor_ that works

Note: Try finding a free custom theme that works and send me that. Simplify 10 is paid, and the authors probably don't like it
if its shared here.

**Comment by chinagreenelvis** [comment]
Here is a ZIP file containing two themes: Aero, and a theme that I've customized myself using msstyleEditor (Simplify 10 was the base.) It's the only custom theme I have on hand that I know was updated specifically to work with 20H2 and the newer Dark/Light modes.

The Aero folder also contains an .msstyles that was saved using msstyleEditor (aero_saved.mssstyles).

[msstyleEditorX64_aero_not_working.zip](https://github.com/nptr/msstyleEditor/files/8068863/msstyleEditorX64_aero_not_working.zip)

Neither the original nor the saved versions of Aero can be previewed using msstyleEditor, and no changes made to a saved version of Aero take effect when applied via a new theme. (All folder paths and file names double-checked when tested.)

Have you had success in using the Aero theme as a base and creating a new style with changes?

**Comment by chinagreenelvis** [comment]
What is the most recent version of Windows 10 you've been able to successfully edit a copy of the aero.msstyles with? I don't recall ever being able to do it with any version, myself, and I've been messing around with msstyleEditor for years now.

It would be great to be able to work from some kind of base without relying on third-party themes.

**Comment by chinagreenelvis** [comment]
It seems that most of the custom themes are created using Windows StyleBuilder; I've created a basic msstyles with no changes saved using that program. It's probably the way to go for creating themes from the ground up, or at least making a basic stylesheet from that program and then using msstyleEdit to change it. Almost seems like they do the same thing, except WSB is able to create a new sheet from whatever hacks they use.

[Base.zip](https://github.com/nptr/msstyleEditor/files/8075093/Base.zip)

https://vsthemes.org/en/software/16338-windows-style-builder.html

**Comment by nptr** [comment]
Does msstyleEditor work with WSB's base style?

**Comment by chinagreenelvis** [comment]
Yep. They've obviously found some way of creating a working stylesheet from the ground up that circumvents whatever limitations are put in place by Aero.

I'm honesty surprised you've been able to edit Aero at all, if you have. From what I'm being told by everyone else it's never been possible.

**Comment by chinagreenelvis** [comment]
At any rate, yours is a great program that I've found extremely useful. So, for what it's worth, thanks.

**Comment by sh1ft-github** [comment]
Hello. I believe I'm having a similar if not the same issue. 

I've been trying to change some image resource files by editing the aero.msstyles at C:\Windows\Resources\Themes\aero using msstyleEditor but couldn't apply the changes and got the OS rejection error which lead me to believe that the issue might possibly be caused by the file being in use, I ended up saving the modified aero.msstyles file and replacing it in a mounted Windows installation ISO in it's original location. On boot I can get into the initial setup but once reaching the part in which you choose languages, keyboard layouts, region etc... I end up with only a black screen and can only see the mouse cursor. I cannot seem to be able to resolve this issue.

I'm fine with not being able to apply the theme as long as I get to successfully replace it in the ISO file and end up with a working system. 

The aero.msstyles file I've edited originates from the exact version and build of Windows 10 1909. However if I take the original unmodified aero.msstyles file from a newer version of Windows, copy & replace it in the older one it actually works just fine which makes me think that it's the file's signature which is screwed somehow. 

I remember reading one of the linked threads which mentioned this. Has this been implemented into the program since or do I have to sign the file myself if possible? You mentioned creating a dummy signature of sort and using that. Could you possibly share it with us somehow or explain how this is actually done? Is there something that I'm missing which is required for it to work. I've also seen convos about UxStyle and the uxstyle.dll file is that still a thing? If yes is simply making a VM, patching the system with UxStyle, extracting the DLL and copying it to the adequate directory enough?

I don't think the issue is directly related to your program since making adjustments to the same image resources via Resource Hacker produces the same output. Also changing the contents of aero.theme file does works just fine and doesn't result in this error.

I would like if you could elaborate on my misunderstanding. I'm fairly new to Windows .theme/.msstyles modification so please be patient with me. Thank you so much for your work so far. Your program would be a life-saver for me if it would get me to finally modify the default aero.msstyles resource which would work with the OS properly.

Thank you for any follow-up on this.

**Comment by nptr** [comment]
@chinagreenelvis WSB is just a newer version of VSB afaik. I removed the link because WSB is paid for and I am not sure the download in question is endorsed by the original authors. In any case, I am trying to fix the issues you mentioned, but it is a slow process as time is scarce atm.

@sh1ft-github You absolutely need to patch UxTheme. What it does, is removing the signature validation so third party themes can be used. The signature block I am appending to the style is, as you correctly understood, just an invalid signature because it looks like even tho signature validation is disabled, the signature block itself still has to be there.

If you are using [UltraUXThemePatcher](https://mhoefs.eu/software_uxtheme.php), copying the patched DLLs (uxtheme.dll, uxinit.dll, themeui.dll according to the website) might work for your custom ISO. You might be able to replace aero.
If you are using [SecureUxTheme](https://github.com/namazso/SecureUxTheme), it won't. This is because the tool works in-memory instead of modifying the system files. At the installation dialog, the tool is obviously not running yet.

But I suggest you don't patch replace aero at all, but setup Win 10 in a VM, patch the system with either tool, **add** new themes, customize it etc. and then use Microsofts `dism /Capture-Image [...]` tool to capture a `.wim` file and use this to create an ISO.

**Comment by sh1ft-github** [comment]
> @chinagreenelvis WSB is just a newer version of VSB afaik. I removed the link because WSB is paid for and I am not sure the download in question is endorsed by the original authors. In any case, I am trying to fix the issues you mentioned, but it is a slow process as time is scarce atm.
> 
> @sh1ft-github You absolutely need to patch UxTheme. What it does, is removing the signature validation so third party themes can be used. The signature block I am appending to the style is, as you correctly understood, just an invalid signature because it looks like even tho signature validation is disabled, the signature block itself still has to be there.
> 
> If you are using [UltraUXThemePatcher](https://mhoefs.eu/software_uxtheme.php), copying the patched DLLs (uxtheme.dll, uxinit.dll, themeui.dll according to the website) might work for your custom ISO. You might be able to replace aero.
> If you are using [SecureUxTheme](https://github.com/namazso/SecureUxTheme), it won't. This is because the tool works in-memory instead of modifying the system files. At the installation dialog, the tool is obviously not running yet.
> 
> But I suggest you don't patch replace aero at all, but setup Win 10 in a VM, patch the system with either tool, **add** new themes, customize it etc. and then use Microsofts `dism /Capture-Image [...]` tool to capture a `.wim` file and use this to create an ISO. You always want aero as fallback.

Thanks for the follow-up. I will attempt to use the patched DLLs from UxThemePatcher.

I also understand that your time schedule is tight. Thank you for your work so far.

**Comment by nptr** [comment]
I found the issue yesterday! Next version should have it patched. 

The problem is, that some styles, including aero, have the StringTable in the .mui (multilingual UI) files instead of the language-neutral .msstyles () file. This is to make the visual styles multilingual (for example to apply different fonts for Japanese).

When I save the style, I do that in the temp folder, away from the .mui in their language folders (en-US, ja-JP,..). When applying the style, the theme engine can't find the fonts an thus complains. I'll probably embed the StringTable in the style for now.

**Comment by chinagreenelvis** [comment]
Oh, that's awesome, glad to hear we'll be able to edit using Aero as a base now.

I probably won't be using Windows 11 until it's absolutely necessary, but since the visual styles occasionally become incompatible with new versions, it could be useful to have a way to take a copy of the latest version of Aero and apply applicable styles from an outdated stylesheet to it.

That could also come in handy for self-updating abandoned styles from older version of Windows 10. There are a few of those I'd love to be able to use again.

**Comment by GithubUser96** [comment]
> I found the issue yesterday! Next version should have it patched. 
> 
> The problem is, that some styles, including aero, have the StringTable in the .mui (multilingual UI) files instead of the language-neutral .msstyles () file. This is to make the visual styles multilingual (for example to apply different fonts for Japanese).
> 
> When I save the style, I do that in the temp folder, away from the .mui in their language folders (en-US, ja-JP,..). When applying the style, the theme engine can't find the fonts an thus complains. I'll probably embed the StringTable in the style for now.

Nice!

**Comment by OrthodoxWindows** [comment]
This is great ! I will be able to use Aero Lite as a basis of contruction now.

**Comment by nptr** [comment]
I released [V 2.0.0.0 as alpha version](https://github.com/nptr/msstyleEditor/releases/tag/2.0.0.0-alpha). It contains the fix for this issue. Since this is an alpha, there likely are bugs, so beware.

**How to make it work**
1. The style must initially be loaded from its theme folder. The location of the theme folder does not matter.
2. When saving, choose the _"Save as Standalone"_ option to properly reload the style again.

If the _"Save as Standalone"_ option does not appear, step 1. wasn't done or there is a problem.

**Why this is the case**

* Windows needs the StringTable resource to identify the fonts it shall use
* Aero has the StringTable in its .mui files, not in the .msstyles itself.
* For Windows (and msstyleEditor) to properly load the StringTable from the .mui files, the .msstyles file must reside in its regular theme folder and have a matching name. See the graphic below:

```
<my theme folder>
│
├─ en-GB
│   ├─ aero.msstyles.mui
│   └─ aerolite.msstyles.mui
│
├─ <other language folder>
│
├─ Shell
│   └─ NormalColor/...
│
└─ aero.msstyles
```
* To apply a style, msstyleEditor saves it in a temporary location and then tells the OS about it. Since the .msstyles is now removed from its .mui files (wrong folder and mismatching name), the OS can't find the .mui (with its StringTable) and complains with 0x90070490.
=> This was the issue.

* To make it work, msstyleEditor saves the StringTable(s) inside the .msstyles itself and removes the references to the .mui file.
=> This is the reason, why a style must initially be loaded from its theme folder; so msstyleEditor can find the StringTable and relocate it.
  * The StringTable is automatically relocated when using the "Test" feature.
  * The StringTable is relocated when using the _"Save as Standalone"_ option.
  * The "Save" option is minimally invasive and doesn't relocate, so the style only works **for Windows** from inside the theme folder.

**To Do List**
* Make _"Save as Standalone"_ the default option if it proves to be working
* Write documentation
* Hide the technical things from the user to reduce confusion

**Comment by nptr** [comment]
[Version 2.0.0.0](https://github.com/nptr/msstyleEditor/releases/tag/2.0.0.0) is out of beta and the most important thing from the To-Do List from above, namely _"Make "Save as Standalone" the default option"_ , is done. The rest comes gradually. I'll consider this issue now fixed.
